### DISCUSSION TITLE
Getting a `Color` from `HSL`

### DISCUSSION THREAD

**Discussion by Atulin** [SixLabors/ImageSharp#2044] [Ideas] [answered]
There's no `Help` category, so let's just say I'll treat the `Ideas` one as `I have no idea`.

I want to create a gradient between two HSL colours, and I'm completely stuck at doing it. There seems to be no existing way to create a colour from HSL, and `ColorStop` requires an instance of `Color`.

There's no `Color.FromHsl()`, no `Color()` ctor that would take HSL, no `Hsl.ToRgba32()`, no nothing. I had to double-check if maybe `Hsl` struct comes from some different library, but no, it does not.

**Comment by Atulin** [comment, marked as answer]
And immediately as I wrote it, the issue is solved...

```cs
var color = new Hsl(hue, saturation, lightness);
var converter = new ColorSpaceConverter();
var c = converter.ToRgb(color);
var stop = new ColorStop(0f, new Color((Rgba32)c));
```

never thought I'd need so much code to do it, but at least it works.

**Reply by JimBobSquarePants** [reply]
Could be worse, at least you didn’t have to write an entire 2D graphics API from scratch that people take for granted and treat the owners with disrespect

**Reply by Atulin** [reply]
Indeed, could be worse. And a quick extension method simplified the whole thing into being a non-issue.

**Reply by JimBobSquarePants** [reply]
>There's no Help category, so let's just say I'll treat the Ideas one as I have no idea.

I want to highlight here that you are misusing the Ideas Discussion forum for questions.

![image](https://user-images.githubusercontent.com/385879/157134591-47653bce-aced-411c-8a2c-1b405fa649e3.png)

You don't come to this forum, misuse it, and complain about how many lines of code you had to write. If you do so again you will be blocked. 

There is no help category by design. This software is released [AS IS](https://github.com/SixLabors/ImageSharp/blob/1859d0f1bcb012083ab6e55386bd9e3c0b291e2f/LICENSE#L143-L151) unless you [purchase a Six Labors License](https://sixlabors.com/pricing/).

**Reply by Atulin** [reply]
Ah, gotcha, my bad. Next time I'll try stumbling through the barebones documentation a bit more 👌

**Reply by JimBobSquarePants** [reply]
Aaaaand that’s a block

**Reply by LucasVos** [reply]
@JimBobSquarePants  First of all: compliments on the lib! It is a bliss to work with a truely .net image library. Especially since System.Drawing was awkward and non-compatible with any Azure stuff. So please, don't block me :-) 

But... I must say.. I feel some sympathy on @Atulin because I stumbled on the same problem...

I wanted to know how to go from a ColorSpace HSL color to a Rgba32, simply to manipulate a simple pixel while using your advised Row Based pixel manipulation.
And searched and searched in the documentation, but could not find a decent solution. And so, i must say i was also surprised to find out that colorspace colors are not easily converted to the pixel domain of an image and it took me also 4 lines of code to do the conversion. @Atulin had a point here... Please let me know if i need to clear  up things!

**Reply by br3aker** [reply]
> And searched and searched in the documentation, but could not find a decent solution.

I want to intervene here. Don't get me wrong but you always can:
1. Contact owners to help them extend documentation
2. Propose a better solution in `Ideas` forum
3. Actually implement your proposed solution to help ImageSharp become a better library

"This library is bad because I HAD to do some researcsh and DOUBLE CHECK search bar on github" won't magically help anyone. He didn't even state what exactly is wrong, 4 lines too much? Lack of documentation? I beg your pardon, this is nothing but a waste of time which could have been spent on actual development.

**Atulin** is not banned from nugget so he can still use ImageSharp, the only thing he lost access to - ability to develop ImageSharp which he clearly didn't want to utilize anyway (and he still can create a new account).

**Reply by LucasVos** [reply]
@br3aker You are right.. That would have been the better approach for Atulin.
And I could have clarified a bit more on the suggestion. I really don't wont to offend anyone and certainly did not state that this library is bad, the **opposite**.

I stumbled upon this library because it is used in Microsoft documentation regarding Azure Functions, in a demo for resizing images. What a great achievement that your code is used in their documentation. _Be proud!_ This is a truly amazing library in .NET.

I feel a bit reluctant to clarify myself now, because i'm a little bit afraid for reactions, but i will try...

I will simply reflect my own first time experience with the library. 
Which do not state that anything is wrong or bad, but it is just my experience. 

First time users are the ones who typically are 'consumers' and not 'contributors'. They can become 'contributors' after their experience as a 'consumer'. Consumer only ask 'stupid' questions or even put remarks with bad choice of words. I suggest that you don't expect them to be contributor directly. There comments could in fact be useful data, if you want to support first-time (consumer) users more. 

Suggestion
==========
As a new/fresh user of the library:

- I had trouble to understand how to use ColorSpaces during pixel manipulation. (= my problem)
- As a new-user, i would expect that HSL color, LAB colors can 'easily' be used during pixel manipulation.  (my expectation) 
- The 4 lines of needed code, simply 'feels' long, because of the emphasize on speed, span<> , etc. during pixel manipulation. (=my feeling) 

(again : these are my thoughts, nothing is complaining about the library here)

As a suggestion you might consider to introduce :
- An additional example of this use case ( doing pixel manipulation with colorspaces)
- Add more method overloads on the ColorSpace objects , that supports a one-liner conversion back to pixel data ?

You can use this information as a measurement, as data, as how your library is perceived on first time use. A consumer that gives feedback.  

Just _be proud_ of your work anyhow.
Dont get offended by the 'consumers'. 
It is in your complete right to simply ignore their requests and post. 
It is your library.
It is just a experience shared with, some data point, some information you can use to make further decision on your roadmap.. 
And yes, consumers should be kind, ask questions. That's true! :-) 
Consumers are not Contributors directly.


On meta level 
=============
I think, as the users of this library grows, you will see these two types of users more and more. The flyby/consumers/i-want-it-now developers and the ones who actually contribute something.

The consumer guys won't take the time to read, etc.. They are the quick-win users, want their answers directly. These type of users ask the stupid questions, complain about their missing 'thingie'...What a pain in the ass!! Yes.. But... I think that is natural. Again, dont get offended by them, just _be proud_ of your work! 

I suppose that you just read the comments of the consumer guys, decide if there is some usefulness in their comments and decide if you want to support these group of library users, their cases, etc. Otherwise , simply ignore them. 

And it is true that you invest so much energy and time. Simply be bigger then the 'consumers' and look at them with a different perspective. _Be proud_ regardless of their comments.

Just consider them as a second group of 'users' and decide if you want to support them by excessive samples, demo code, documentation of all kinds of use cases, etc. And again, just _be proud_ that you've actually managed to attract the consumer users.. because it means that your lib is gaining attraction more and more. What an achievement! 

Keep up the very good work! 
I probably will use the library more, and gradually become a Contributor instead of a Consumer. :-)

**Reply by JimBobSquarePants** [reply]
Here's the thing @LucasVos and I'll be brief. 

I understand you do not mean to offend anyone but you are forgetting something fundamental when you describe Consumers vs Contributors.

ImageSharp is licenced AS-IS. This means that there is absolutely zero obligation for anyone here to answer questions or even fix issues. You use this AT YOUR OWN RISK. If as a Consumer you (figuratively) are not happy with that we off an alternative commercial license.
 
The Contributors to this library are already pretty proud of it and usage is already massive. We have over 24 million downloads on NuGet and know for damn sure we have the best library in the entire .NET ecosystem.  

So please try to consider things from this perspective. 

Converting from HSL or LAB is relatively simple. You can do it in 4 lines of code and you can do it because someone spent literally months working evenings after they had put their child to bed and said goodnight to their wife, often working until 2AM. All while maintaining a full time job unrelated to this library. 

Not only can you convert those color spaces, you can convert an additional 12 different color spaces using the same code! You can convert individual pixels or entire rows! You can do it for free! Not bad eh?

We gladly except feature requests when they are delivered in a format that is constructive.

Atulin was arrogant and rude and came to this discussion with an attitude that is entirely unwelcome. He had no intention in contributing and was blocked for his behavior. 

I take a very firm stance in this regard because I have been maintaining Open Source libraries for over a decade and know fine well the consequences of pandering to the whims of the entitled.
